rholang/src/test/scala/coop/rchain/rholang/interpreter/error/MLockSpec.scala

```diff
@@ -0,0 +1,62 @@
+package coop.rchain.rholang.interpreter.error
+
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import monix.eval.Task
+import org.scalatest.{FlatSpec, Matchers}
+
+import scala.concurrent.duration._
+
+class MLockSpec extends FlatSpec with Matchers {
+
+  import monix.execution.Scheduler.Implicits.global
+
+  private val MAX_DURATION = 10.seconds
+
+  def withFixture[A](f: MLock[Task] => Task[A]): A =
+    (MLock[Task] >>= f).runSyncUnsafe(MAX_DURATION)
+
+  behavior of "MLock"
+
+  "release" should "throw error when released lock is released" in withFixture(
+    _.release.attempt
+      .map(_ should be(Left(LockReleaseException("attempted to release released lock"))))
+  )
+
+  "release" should "unblock lock acquisitions in FIFO order" in withFixture { lock =>
+    for {
+      ref <- Ref.of[Task, Int](0)
+      _   <- lock.acquire
+      result <- List(
+                 lock.acquire >> ref.getAndSet(1),
+                 Task.sleep(0.1.seconds) >> lock.acquire >> ref.getAndSet(2),
+                 Task.sleep(0.2.seconds) >> lock.release >> Task.sleep(0.1.seconds) >> ref.get,
+                 Task.sleep(0.4.seconds) >> lock.release >> Task.sleep(0.1.seconds) >> ref.get
+               ).parSequence
+    } yield assert(result.head == 0 && result(1) == 1 && result(2) == 1 && result(3) == 2)
+  }
+
+  "withLock" should "block lock acquisitions while lock is held" in withFixture { lock =>
+    for {
+      ref <- Ref.of[Task, Int](0)
+      result <- List(
+                 lock.withLock(Task.sleep(0.3.seconds) >> ref.get),
+                 Task.sleep(0.1.seconds) >> lock.acquire >> ref.getAndSet(1),
+                 Task.sleep(0.5.seconds) >> ref.get
+               ).parSequence
+    } yield assert(result.head == 0 && result(1) == 0 && result(2) == 1)
+  }
+
+  "release" should "unblock all waiting threads simultaneously" in withFixture { lock =>
+    for {
+      ref   <- Ref.of[Task, Int](0)
+      _     <- lock.acquire
+      count = 999
+      _ <- List(
+            (0 to count).toList.parTraverse_(_ => lock.waitForLock >> ref.update(_ + 1)),
```

Comment:
I'll answer to your comment https://github.com/rchain/rchain/pull/2504#issuecomment-500872011 here as this is the only usage of `lock.waitForLock`. If you remove `lock.waitForLock >>` from the test above it still passes. This is due to the nature of `ref.update` as stated in the docs:
```/**
   * Modifies the current value using the supplied update function. If another modification
   * occurs between the time the current value is read and subsequently updated, the modification
   * is retried using the new value. Hence, `f` may be invoked multiple times.
   *
   * Satisfies:
   *   `r.update(_ => a) == r.set(a)`
   */
  def update(f: A => A): F[Unit]
```

Maybe the functionality of `waitForLock` isn't absolutely necessary?

Reply:
Its functionality is required in a following PR:

```
final class RWLock[F[_]](stateRef: MVar[F, State], readLock: MLock[F], writeLock: MLock[F])(
    implicit bracketF: Bracket[F, Throwable]
) {

  /**
    * Acquires read lock.
    *
    * Blocks (asynchronously) if another thread has write lock.
    *
    */
  def acquireRead: F[Unit] =
    stateRef.take >>= {
      case Free        => readLock.acquire >> stateRef.put(Read(1))
      case Write       => stateRef.put(Write) >> writeLock.waitForLock >> acquireRead
      case Read(count) => stateRef.put(Read(count + 1))
    }
...
}